File: sdk/python/agentmesh/_span.py
"""AgentMesh SDK's span domain model.

Mirrors the shape defined in shared/span/span.go and docs/otlp-mapping.md.
This is a deliberate parallel implementation, not a shared import — the Go
`shared/span` package and this Python module are separate language
ecosystems that agree only through the wire contract in
docs/otlp-mapping.md, exactly as that document states: "This document is
the wire contract between every OTLP exporter ... and the Collector's
decoder." Any change to the span model's meaning must be made in both
places and reflected in that document.
"""

from __future__ import annotations

import enum
import os
import time
from dataclasses import dataclass, field

# Mirrors shared/span.CurrentSchemaVersion. Any wire-format change bumps
# both this value and the Go constant simultaneously (docs/otlp-mapping.md).
CURRENT_SCHEMA_VERSION = 1


class SpanKind(str, enum.Enum):
    """Mirrors shared/span.Kind (Architecture.md §3)."""

    LLM_CALL = "llm.call"
    TOOL_CALL = "tool.call"
    AGENT_HANDOFF = "agent.handoff"
    MCP_CALL = "mcp.call"


class SpanStatus(str, enum.Enum):
    """Mirrors shared/span.Status."""

    OK = "ok"
    ERROR = "error"
    TIMEOUT = "timeout"
    DENIED = "denied"


# Per docs/otlp-mapping.md's "Payload size threshold" section (corrected):
# the 4KB inline/blob-ref decision is made by the Collector on ingestion,
# not by the exporter. The SDK always sends the full payload inline and
# never truncates or uploads to blob storage itself — doing so would
# require distributing object-storage write credentials to every
# customer's agent process, which Architecture.md §2 explicitly rules out
# ("[the SDK] never sees the Trace Store directly").


def _new_trace_id() -> str:
    """Generate a 32-hex-char trace id, matching shared/ids.TraceID's wire
    format (a 16-byte value rendered as lowercase hex)."""
    return os.urandom(16).hex()


def _new_span_id() -> str:
    """Generate a 16-hex-char span id, matching shared/ids.SpanID's wire
    format (an 8-byte value rendered as lowercase hex)."""
    return os.urandom(8).hex()


@dataclass
class Span:
    """A single unit of work inside a trace, prior to OTLP encoding.

    Mirrors shared/span.Span's fields that are populated client-side; the
    Collector-computed fields (schema validation results, ingested_at,
    blob_ref) have no client-side representation — the SDK never decides
    where a payload is stored, only what it is (docs/otlp-mapping.md's
    corrected "Payload size threshold" section).
    """

    project_id: str
    kind: SpanKind
    name: str
    trace_id: str = field(default_factory=_new_trace_id)
    span_id: str = field(default_factory=_new_span_id)
    parent_span_id: str | None = None
    start_time_ns: int = field(default_factory=time.time_ns)
    end_time_ns: int | None = None
    status: SpanStatus | None = None
    input: str | None = None
    output: str | None = None
    token_input: int | None = None
    token_output: int | None = None
    cost_usd: float | None = None
    attributes: dict[str, str] = field(default_factory=dict)

    def finish(
        self,
        *,
        status: SpanStatus = SpanStatus.OK,
        output: str | None = None,
        token_input: int | None = None,
        token_output: int | None = None,
        cost_usd: float | None = None,
    ) -> None:
        """Mark the span complete. Mirrors setting EndTime/Status/Output on
        the Go side; called once the wrapped LLM/tool call returns."""
        self.end_time_ns = time.time_ns()
        self.status = status
        if output is not None:
            self.output = output
        if token_input is not None:
            self.token_input = token_input
        if token_output is not None:
            self.token_output = token_output
        if cost_usd is not None:
            self.cost_usd = cost_usd

    def set_input(self, input_value: str) -> None:
        self.input = input_value

File: shared/span/span.go
// Package span defines AgentMesh's core domain model: the Span and Trace
// types that every service (Collector, Query API, Replay Engine, Anomaly
// Detector, Cost Engine) shares.
//
// This is the single Go-native source of truth for the shape defined in
// System Design.md §2.1 (the ClickHouse spans table) and Architecture.md §3
// (the four span kinds and the framework-to-span mapping). Every service
// that reads or writes span data imports this package rather than declaring
// its own struct — the classic monorepo failure mode this guards against is
// two services silently drifting on what a field means (Repository
// Structure.md's rationale for keeping schema definitions out of any single
// service's ownership).
//
// SchemaVersion is embedded on every Span so a Collector or SDK version
// mismatch during a rolling upgrade is detectable rather than silently
// producing malformed data (Technical Roadmap.md §9).
package span

import (
	"fmt"
	"time"

	"github.com/agentmesh/agentmesh/shared/ids"
)

// CurrentSchemaVersion is the schema version this build of the span package
// implements. Bump it whenever the Span struct's wire-relevant shape changes
// in a way older consumers cannot safely interpret.
const CurrentSchemaVersion = 1

// Kind identifies what a Span represents, per Architecture.md §3.
type Kind string

const (
	// KindLLMCall represents a single LLM request/response.
	KindLLMCall Kind = "llm.call"
	// KindToolCall represents a single tool/function invocation.
	KindToolCall Kind = "tool.call"
	// KindAgentHandoff represents control passed from one agent/sub-agent to
	// another.
	KindAgentHandoff Kind = "agent.handoff"
	// KindMCPCall represents a tool call routed through the MCP Gateway,
	// captured automatically even if the calling agent has no SDK
	// integration.
	KindMCPCall Kind = "mcp.call"
)

// ValidKinds lists every recognized Kind, used for validation.
var ValidKinds = map[Kind]bool{
	KindLLMCall:      true,
	KindToolCall:     true,
	KindAgentHandoff: true,
	KindMCPCall:      true,
}

// Status is the terminal outcome of a Span.
type Status string

const (
	StatusOK      Status = "ok"
	StatusError   Status = "error"
	StatusTimeout Status = "timeout"
	// StatusDenied is emitted by the MCP Gateway when a guardrail policy
	// rejects a call (System Design.md §6).
	StatusDenied Status = "denied"
)

// ValidStatuses lists every recognized Status, used for validation.
var ValidStatuses = map[Status]bool{
	StatusOK:      true,
	StatusError:   true,
	StatusTimeout: true,
	StatusDenied:  true,
}

// Payload holds a span's input or output value, either inline (small
// payloads, System Design.md §2.1's 4KB threshold) or as a reference to a
// blob-store object (large payloads, Architecture.md §14). Exactly one of
// Inline or BlobRef is set; never both, per InlineOrBlobRef's invariant.
type Payload struct {
	Inline  string
	BlobRef string
}

// IsInline reports whether the payload is stored inline rather than in blob
// storage.
func (p Payload) IsInline() bool { return p.BlobRef == "" }

// IsEmpty reports whether no payload was recorded at all.
func (p Payload) IsEmpty() bool { return p.Inline == "" && p.BlobRef == "" }

// Span is one unit of work inside a Trace (System Design.md §2.1).
type Span struct {
	SchemaVersion int
	ProjectID     ids.ProjectID
	TraceID       ids.TraceID
	SpanID        ids.SpanID
	ParentSpanID  ids.SpanID // zero value means "root span"

	Kind      Kind
	Name      string // model name, tool name, or agent name depending on Kind
	StartTime time.Time
	EndTime   time.Time
	Status    Status

	Input  Payload
	Output Payload

	TokenInput  *uint32 // nil means "not applicable" (e.g., tool.call spans)
	TokenOutput *uint32
	CostUSD     *float64 // nil means "cost unknown", never assumed to be zero (System Design.md §7)

	// Attributes is the open-ended key/value bag (e.g., framework name,
	// model version) from System Design.md §2.1's `attributes Map(String,
	// String)` column.
	Attributes map[string]string

	// ChaosInjected and ChaosFaultType surface Phase 2's chaos-engineering
	// feature (sdk/python/agentmesh/chaos.py): whether this span's outcome
	// was a deliberately injected fault rather than a natural one, and
	// which kind ("latency" | "error"). Backed by
	// schema/clickhouse/002_chaos_columns.sql's dedicated columns rather
	// than the generic Attributes map, so "every span where a fault was
	// injected" is a direct, indexable column query. ChaosFaultType is
	// empty when ChaosInjected is false.
	ChaosInjected  bool
	ChaosFaultType string
}

// HasParent reports whether this Span is not a trace root.
func (s Span) HasParent() bool { return !s.ParentSpanID.IsZero() }

// Duration returns the span's wall-clock duration.
func (s Span) Duration() time.Duration { return s.EndTime.Sub(s.StartTime) }

// Validate checks the Span's structural invariants: required identifiers are
// set, Kind and Status are recognized values, EndTime is not before
// StartTime, and a Payload never sets both Inline and BlobRef. Validate is
// called by every ingestion path (Collector) before a Span is persisted, so
// malformed data is rejected at the boundary rather than corrupting the
// trace store (Architecture.md §17's error-handling philosophy).
func (s Span) Validate() error {
	if s.ProjectID == (ids.ProjectID{}) {
		return fmt.Errorf("span: ProjectID is required")
	}
	if s.TraceID.IsZero() {
		return fmt.Errorf("span: TraceID is required")
	}
	if s.SpanID.IsZero() {
		return fmt.Errorf("span: SpanID is required")
	}
	if !ValidKinds[s.Kind] {
		return fmt.Errorf("span: unrecognized kind %q", s.Kind)
	}
	if s.Name == "" {
		return fmt.Errorf("span: Name is required")
	}
	if s.Status != "" && !ValidStatuses[s.Status] {
		return fmt.Errorf("span: unrecognized status %q", s.Status)
	}
	if !s.EndTime.IsZero() && s.EndTime.Before(s.StartTime) {
		return fmt.Errorf("span: EndTime (%s) is before StartTime (%s)", s.EndTime, s.StartTime)
	}
	if !s.Input.IsEmpty() && s.Input.Inline != "" && s.Input.BlobRef != "" {
		return fmt.Errorf("span: Input payload sets both Inline and BlobRef")
	}
	if !s.Output.IsEmpty() && s.Output.Inline != "" && s.Output.BlobRef != "" {
		return fmt.Errorf("span: Output payload sets both Inline and BlobRef")
	}
	if s.SchemaVersion == 0 {
		return fmt.Errorf("span: SchemaVersion is required")
	}
	return nil
}
